utils/staking-miner/src/monitor.rs

```diff
@@ -64,120 +68,245 @@ async fn ensure_no_previous_solution<
+
+		let client = Arc::new(client);
+		let (tx, mut rx) = mpsc::unbounded_channel::<StakingMinerError>();
+
 		loop {
-			log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", sub, unsub);
-			let mut subscription: Subscription<Header> = client
-				.subscribe(&sub, None, &unsub)
-				.await
-				.unwrap();
-
-			while let Some(now) = subscription.next().await? {
-				let hash = now.hash();
-				log::trace!(target: LOG_TARGET, "new event at #{:?} ({:?})", now.number, hash);
-
-				// if the runtime version has changed, terminate.
-				crate::check_versions::<Runtime>(client).await?;
-
-				// we prefer doing this check before fetching anything into a remote-ext.
-				if ensure_signed_phase::<Runtime, Block>(client, hash).await.is_err() {
-					log::debug!(target: LOG_TARGET, "phase closed, not interested in this block at all.");
-					continue;
-				};
+			let at = tokio::select! {
+				maybe_rp = subscription.next() => {
+					match maybe_rp {
+						Some(Ok(r)) => r,
+						// Custom `jsonrpsee` message; should not occur.
+						Some(Err(RpcError::SubscriptionClosed(reason))) => {
+							log::debug!("[rpc]: subscription closed by the server: {:?}, starting a new one", reason);
+							continue;
+						}
+						Some(Err(e)) => {
+							log::error!("[rpc]: subscription failed to decode Header {:?}, this is bug please file an issue", e);
+							return Err(e.into());
+						}
+						// The subscription was dropped, should only happen if:
+						//	- the connection was closed.
+						//	- the subscription could not need keep up with the server.
+						None => {
+							log::warn!("[rpc]: restarting header subscription");
+							subscription = client.subscribe(&sub, None, &unsub).await?;
+							log::warn!(target: LOG_TARGET, "subscription to {} terminated. Retrying..", sub);
+							continue
+						}
+					}
+				},
+				maybe_err = rx.recv() => {
+					match maybe_err {
+						Some(err) => return Err(err),
+						None => unreachable!("at least one sender kept in the main loop should always return Some; qed"),
+					}
+				}
+			};
+
+			log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?} at: {}", sub, unsub, at.number());
+
+			// Spawn task and non-recoverable errors are sent back to the main task
+			// such as if the connection has been closed.
+			tokio::spawn(
+				send_and_watch_extrinsic(client.clone(), tx.clone(), at, signer.clone(), shared.clone(), config.clone())
+			);
+		}
+
+		/// Construct extrinsic at given block and watch it.
+		async fn send_and_watch_extrinsic(
+			client: Arc<WsClient>,
+			tx: mpsc::UnboundedSender<StakingMinerError>,
+			at: Header,
+			signer: Signer,
+			shared: SharedConfig,
+			config: MonitorConfig,
+		) {
+
+			let hash = at.hash();
+			log::trace!(target: LOG_TARGET, "new event at #{:?} ({:?})", at.number, hash);
+
+			// if the runtime version has changed, terminate.
+			if let Err(err) = crate::check_versions::<Runtime>(&*client).await {
+				let _ = tx.send(err.into());
```

Comment:
Just to confirm my understanding, this will send an error back to the thread that spawned this and that will be caught by the above line `maybe_err = rx.recv()`?

Reply:
yeah, exactly.